### PR TITLE
Add generation of other.xml.gz file

### PR DESCRIPTION
**Pull request by VitaliyaIoffe**
Mkrepo didn't generate other.xml.gz, which is needed for adding,
for example, CentOs 7. Additional methods were added to a script
rpmrepo.py, so that it would generate other.xml.gz file, as there
was only two files filelists and primary .xml.gz with metainformation
previously.

repodata/other.xml.gz, with the description this file currently
only stores the changelog data from packages.
http://blog.sweetxml.org/2008/03/anatomy-of-yum-repository-look-under.html
https://en.opensuse.org/openSUSE:Standards_Rpm_Metadata#How_to_extend_rpm-md_metadata_with_non_standard_data

Was added several unittests to check dump_other and
header_to_other functions.

Close knazarov/mkrepo#32

I think it depends now on the #41. 
I tried creating the other file with the next diff

```
@@ -961,12 +961,20 @@ def update_repo(storage, sign, tempdir, force=False):
         filelists, primary, others, revision = parse_repomd(data)
 
         initial_filelists = filelists['location']
-        data = storage.read_file(initial_filelists)
-        filelists = parse_filelists(gunzip_bytes(data))
+        if storage.exists(initial_filelists):
+            data = storage.read_file(initial_filelists)
+            filelists = parse_filelists(gunzip_bytes(data))
+        else:
+            initial_filelists = None
+            filelists = {}
 
         initial_primary = primary['location']
-        data = storage.read_file(initial_primary)
-        primary = parse_primary(gunzip_bytes(data))
+        if storage.exists(initial_primary):
+            data = storage.read_file(initial_primary)
+            primary = parse_primary(gunzip_bytes(data))
+        else:
+            initial_primary = None
+            primary = {}
 
         if others and storage.exists(initial_other):
             initial_other = others['location']
```
 
Also, I want to do additional checks.

### CODE REVIEW

File: rpmrepo.py [hunk 1]
```diff
@@ -491,7 +548,99 @@ def parse_ver_str(ver_str):
...
+    for package in other.values():
+        res += '<package pkgid="%s" name="%s" arch="%s">\n' % (
+            package['pkgid'], package['name'], package['arch'])
+
+        ver = package['version']
+        log = package['changelog']
+
+        res += '  <version '
+        components = ' '.join(
+            ['%s="%s"' % (c, ver[c]) for c in ['epoch', 'ver', 'rel'] if ver[c]])
```
Comment (VitaliyaIoffe) [inline rpmrepo.py:629]: Maybe it would be good to set preparation before res define

File: rpmrepo.py [hunk 2]
```diff
@@ -377,6 +384,56 @@ def parse_primary(data):
...
 
+def parse_other(data):
+    """Parse other.xml
+
+    Find 'other:changelog' and 'other:version' and
+    fill changelog, version and package.
+
+    Keyword Arguments:
+    data - The unparsed data of other.xml.gz (byte, default: None)
+    Return Parsed data as a packages (dict)
```
Comment (LeonidVas) [inline rpmrepo.py:395]: Usually, an empty line is placed before the `Return` description to separate it from the arguments.

File: rpmrepo.py [hunk 3]
```diff
@@ -491,7 +548,99 @@ def parse_ver_str(ver_str):
...
-def get_with_decode(dictionary, key, default='' , encoding='utf-8'):
+def header_to_other(header, sha256):
+    """Method that decodes data for parsing in sha256
+
+    Keyword arguments:
+    header - data that get passed
+    to decode (string, default: None)
+    sha256 - code format that
+    is assigned for generating package id (string)
+    Return data that contains
```
Comment (LeonidVas) [inline rpmrepo.py:559]: Usually, an empty line is placed before the `Return` description to separate it from the arguments.

File: rpmrepo.py [hunk 4]
```diff
@@ -491,7 +548,99 @@ def parse_ver_str(ver_str):
     return (epoch, ver, rel)
 
 
-def get_with_decode(dictionary, key, default='' , encoding='utf-8'):
+def header_to_other(header, sha256):
+    """Method that decodes data for parsing in sha256
+
+    Keyword arguments:
+    header - data that get passed
+    to decode (string, default: None)
```
Comment (LeonidVas) [inline rpmrepo.py:556]: Move to the previous line.

File: rpmrepo.py [hunk 5]
```diff
@@ -491,7 +548,99 @@ def parse_ver_str(ver_str):
...
+    sha256 - code format that
+    is assigned for generating package id (string)
```
Comment (LeonidVas) [inline rpmrepo.py:558]: Move to the previous line.

File: rpmrepo.py [hunk 6]
```diff
@@ -801,6 +968,14 @@ def update_repo(storage, sign, tempdir, force=False):
         data = storage.read_file(initial_primary)
         primary = parse_primary(gunzip_bytes(data))
 
+        if others and storage.exists(initial_other):
```
Comment (LeonidVas) [inline rpmrepo.py:971]: It doesn't work so)
You must get `initial_other` before to check if it exists.

**Comment by LeonidVas** [comment]
> When I tested this patch on the third run I got the following exception:
> 
> ```
> Updating rpm repository: tarantool-dev/restored/release/1.10/el/7/x86_64
> Traceback (most recent call last):
>   File "/home/leonid/work/mail/mkrepo/./mkrepo.py", line 120, in <module>
>     main()
>   File "/home/leonid/work/mail/mkrepo/./mkrepo.py", line 116, in main
>     update_repo(path, args)
>   File "/home/leonid/work/mail/mkrepo/./mkrepo.py", line 58, in update_repo
>     rpmrepo.update_repo(stor, args.sign, args.temp_dir, args.force)
>   File "/home/leonid/work/mail/mkrepo/rpmrepo.py", line 1002, in update_repo
>     initial_filelists, initial_primary, initial_other) = parse_metafiles(storage)
>   File "/home/leonid/work/mail/mkrepo/rpmrepo.py", line 991, in parse_metafiles
>     others = parse_other(gunzip_bytes(data))
>   File "/home/leonid/work/mail/mkrepo/rpmrepo.py", line 398, in parse_other
>     root = ET.fromstring(data)
>   File "/usr/lib/python3.9/xml/etree/ElementTree.py", line 1347, in XML
>     parser.feed(text)
> xml.etree.ElementTree.ParseError: not well-formed (invalid token): line 5, column 34
> ```

I added a commit to fix this bug. If you are agree with me squash this changes to your commit.

**Review by LeonidVas — APPROVED** [review]
LGTM.

**Comment by LeonidVas** [comment]
> > When I tested this patch on the third run I got the following exception:
> > ```
> > Updating rpm repository: tarantool-dev/restored/release/1.10/el/7/x86_64
> > Traceback (most recent call last):
> >   File "/home/leonid/work/mail/mkrepo/./mkrepo.py", line 120, in <module>
> >     main()
> >   File "/home/leonid/work/mail/mkrepo/./mkrepo.py", line 116, in main
> >     update_repo(path, args)
> >   File "/home/leonid/work/mail/mkrepo/./mkrepo.py", line 58, in update_repo
> >     rpmrepo.update_repo(stor, args.sign, args.temp_dir, args.force)
> >   File "/home/leonid/work/mail/mkrepo/rpmrepo.py", line 1002, in update_repo
> >     initial_filelists, initial_primary, initial_other) = parse_metafiles(storage)
> >   File "/home/leonid/work/mail/mkrepo/rpmrepo.py", line 991, in parse_metafiles
> >     others = parse_other(gunzip_bytes(data))
> >   File "/home/leonid/work/mail/mkrepo/rpmrepo.py", line 398, in parse_other
> >     root = ET.fromstring(data)
> >   File "/usr/lib/python3.9/xml/etree/ElementTree.py", line 1347, in XML
> >     parser.feed(text)
> > xml.etree.ElementTree.ParseError: not well-formed (invalid token): line 5, column 34
> > ```
> 
> I added a commit to fix this bug. If you are agree with me squash this changes to your commit.

The commit has been squashed.

File: rpmrepo.py [hunk 7]
```diff
@@ -491,7 +549,100 @@ def parse_ver_str(ver_str):
...
+    if not isinstance(changelog_text, list):
+        changelog_text = [changelog_text]
+    changelog_date = header.get('CHANGELOGTIME', [])
+    if not isinstance(changelog_date, list):
+        changelog_date = [changelog_date]
+
+    # get last CHANGELOG_LIMIT lines of changelogs in reverse order
+    changelog_name = changelog_name[:CHANGELOG_LIMIT][::-1]
+    changelog_text = changelog_text[:CHANGELOG_LIMIT][::-1]
+    changelog_date = changelog_date[:CHANGELOG_LIMIT][::-1]
```
Comment (Totktonada) [inline rpmrepo.py:590]: AFAIS, here we get **first** CHANGELOG_LIMIT items in the reverse order.

```python
>>> ['a', 'b', 'c', 'd', 'e', 'f'][:3][::-1]
['c', 'b', 'a']
```

**Review by Totktonada — APPROVED** [review]
Visually it looks correct (and I hope you test it manually on real packages).

I have only one question and ready to merge after response / fixup.

File: rpmrepo.py (hunk 7 above)
Comment (VitaliyaIoffe) [reply rpmrepo.py]: Add the comment which describes that we get the newest changelogs

Comment (Totktonada) [reply rpmrepo.py]: I'll cite the new version:

```python
# get last CHANGELOG_LIMIT lines of changelogs in reverse order
# the first line is an initial commit, we get the newest lines in
# terms of time, we ordered these lines as a creatrepo does
# so, reverse them
changelog_name = changelog_name[:CHANGELOG_LIMIT][::-1]
changelog_text = changelog_text[:CHANGELOG_LIMIT][::-1]
changelog_date = changelog_date[:CHANGELOG_LIMIT][::-1]
```

It still looks controversial for me. Let's assume we have the following list (sorted from oldest to newest here):

* one
* two
* ...
* eleven
* twelve

Next, read your comment:

> the first line is an initial commit

I guess it is the description of the input data (what we get from headers). Okay, so it is actually follows as shown above in the input data.

Next, what we want to achieve:

> we get the newest lines in terms of time

I guess newest CHANGELOG_LIMIT (`10`) ones (disregarding order):

* three
* four
* ...
* eleven
* twelve

Let's try the code:

```python
>>> CHANGELOG_LIMIT = 10
>>> ['one', 'two', 'three', 'four', 'five', 'six', 'seven', 'eight', 'nine', 'ten', 'eleven', 'twelve'][:CHANGELOG_LIMIT]
['one', 'two', 'three', 'four', 'five', 'six', 'seven', 'eight', 'nine', 'ten']
```

Nope, it does not work in the way as described: we see `'one'` and `'two'`, but don't see `'eleven'` and `'twelve'`. We got oldest entries, not newest. Or the description allows several ways to interpret it.

Do you get what I don't understand here?

**Pull request merged**